assets/js/base/hooks/use-store-cart-item.js

```diff
@@ -22,32 +24,60 @@ import { useStoreCart } from './use-store-cart';
  */
 export const useStoreCartItem = ( cartItemKey ) => {
 	const { cartItems, cartIsLoading } = useStoreCart();
-	const cartItem = cartItems.filter( ( item ) => item.key === cartItemKey );
-
-	const results = useSelect(
-		( select, { dispatch } ) => {
+	const [ cartItem, setCartItem ] = useState( {
+		key: '',
+		isLoading: true,
+		cartData: {},
+		quantity: 0,
+		isPending: false,
+		changeQuantity: () => void null,
+		removeItem: () => void null,
+	} );
+	// Store quantity in hook state. This is used to keep the UI
+	// updated while server request is updated.
+	const [ quantity, changeQuantity ] = useState( cartItem.quantity );
+	const [ debouncedQuantity ] = useDebounce( quantity, 400 );
+	const isPending = useSelect(
+		( select ) => {
 			const store = select( storeKey );
-			const isPending = store.isItemQuantityPending( cartItemKey );
-			const { removeItemFromCart, changeCartItemQuantity } = dispatch(
-				storeKey
-			);
-
-			return {
-				isPending,
-				changeQuantity: ( newQuantity ) => {
-					changeCartItemQuantity( cartItemKey, newQuantity );
-				},
-				removeItem: () => {
-					removeItemFromCart( cartItemKey );
-				},
-			};
+			return store.isItemQuantityPending( cartItemKey );
 		},
 		[ cartItemKey ]
 	);
+	useEffect( () => {
+		if ( ! cartIsLoading ) {
+			const foundCartItem = cartItems.find(
+				( item ) => item.key === cartItemKey
+			);
+			if ( foundCartItem ) {
+				setCartItem( foundCartItem );
+			}
+		}
+	}, [ cartItems, cartIsLoading, cartItemKey ] );
+
+	const { removeItemFromCart, changeCartItemQuantity } = useDispatch(
+		storeKey
+	);
+	const removeItem = () => {
+		removeItemFromCart( cartItemKey );
+	};
+
+	// Observe debounced quantity value, fire action to update server when it
+	// changes.
+	useEffect( () => {
+		if ( debouncedQuantity === 0 ) {
+			changeQuantity( cartItem.quantity );
+			return;
+		}
```

Comment:
This covers initializing things when debouncedQuantity is 0 we want to make sure that the quantities update when we've found a cartItem.

Reply:
This causes an extra POST request for each cart item on page load, redundantly re-setting the quantity. Are we ok with that?

Reply:
To close the loop, we definitely did _not_ want an extra POST request. But the issue was that the action (below these lines) was posting even if the quantity was the same as what was in the state. So I fixed this by ensuring the action only pinged the server to update the quantity if needed (by checking the current quantity in the state to make sure it didn't match).